src/utils/stats_alloc/src/task_local.rs

```diff
@@ -0,0 +1,103 @@
+use std::alloc::{GlobalAlloc, Layout};
+use std::ptr::NonNull;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use tokio::task_local;
+
+pub struct TaskLocalAllocator;
+
+static GLOBAL_ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
+
+#[repr(transparent)]
+pub struct TaskLocalBytesAllocated(NonNull<AtomicUsize>);
+
+impl Default for TaskLocalBytesAllocated {
+    fn default() -> Self {
+        unsafe {
+            TaskLocalBytesAllocated(NonNull::new_unchecked(Box::leak(Box::new(
+                AtomicUsize::new(0),
+            ))))
+        }
+    }
+}
+
+unsafe impl Send for TaskLocalBytesAllocated {}
+
+impl TaskLocalBytesAllocated {
+    pub fn val(&self) -> usize {
+        unsafe { self.0.as_ref().load(Ordering::Relaxed) }
+    }
+}
+
+task_local! {
+    pub static BYTES_ALLOCATED: TaskLocalBytesAllocated;
+}
+
+struct TaskLocalAlloc;
+
+unsafe impl GlobalAlloc for TaskLocalAlloc {
+    unsafe fn alloc(&self, layout: std::alloc::Layout) -> *mut u8 {
+        let new_layout =
+            Layout::from_size_align_unchecked(layout.size() + usize::BITS as usize, layout.align());
+        BYTES_ALLOCATED
+            .try_with(|bytes| {
+                bytes.0.as_ref().fetch_add(layout.size(), Ordering::Relaxed);
+                let ptr = GLOBAL_ALLOC.alloc(new_layout);
+                *(ptr as *mut usize) = bytes.0.as_ptr() as usize;
+                let ptr = ptr.add(usize::BITS as usize);
+                ptr
+            })
+            .unwrap_or_else(|_| {
+                let ptr = GLOBAL_ALLOC.alloc(new_layout);
+                *(ptr as *mut usize) = 0;
+                let ptr = ptr.add(usize::BITS as usize);
+                ptr
+            })
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
+        let new_layout =
+            Layout::from_size_align_unchecked(layout.size() + usize::BITS as usize, layout.align());
+        let ptr = ptr.sub(usize::BITS as usize);
+        let bytes = ptr as *const AtomicUsize;
```

Comment:
Is dangling pointer possible?

Reply:
We've refactored the implementation and it'll check whether the stats is valid before mutating.